multicluster/controllers/multicluster/internal/remote_cluster.go

```diff
@@ -0,0 +1,300 @@
+	}
+	config.BearerToken = string(token)
+	config.CAData = crtData
+	mgr, err := ctrl.NewManager(config, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "0",
+		Logger:             log.WithName("controller-runtime"),
+	})
+	if err != nil {
+		log.V(1).Error(err, "unable to create new manager")
+		return nil, err
+	}
+
+	remoteClient, e := client.New(config, client.Options{Scheme: scheme})
+	if e != nil {
+		return nil, e
+	}
+	cluster := &remoteCluster{
+		Client:               remoteClient,
+		log:                  log,
+		ClusterManager:       mgr,
+		ClusterSetID:         clusterSetID,
+		ClusterID:            clusterID,
+		config:               config,
+		scheme:               scheme,
+		Namespace:            clusterSetNamespace,
+		connected:            false,
+		isElectedLeader:      false,
+		remoteClusterManager: remoteClusterManager,
+	}
+
+	err = remoteClusterManager.AddRemoteCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return cluster, nil
+}
+
+func getSecretCACrtAndToken(namespace string, secretName string) ([]byte, []byte, error) {
+	var err error
+	mutex.Lock()
+	if secretCient == nil {
+		secretCient, err = client.New(ctrl.GetConfigOrDie(), client.Options{})
+		if err != nil {
+			mutex.Unlock()
+			return nil, nil, err
+		}
+	}
+	mutex.Unlock()
+	secretObj := &v1.Secret{}
+	secretNamespacedName := types.NamespacedName{
+		Namespace: namespace,
+		Name:      secretName,
+	}
+	err = secretCient.Get(context.TODO(), secretNamespacedName, secretObj)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	caData, found := secretObj.Data[v1.ServiceAccountRootCAKey]
+	if !found {
+		return nil, nil, fmt.Errorf("ca.crt data not found in secret %v", secretName)
+	} else {
+	}
+
+	token, found := secretObj.Data[v1.ServiceAccountTokenKey]
+	if !found {
+		return nil, nil, fmt.Errorf("token not found in secret %v", secretName)
+	}
+
+	return caData, token, nil
+}
+
+func (r *remoteCluster) SendMemberAnnounce() error {
+	log := r.log.WithName("MemberAnnounce" + string(r.ClusterID))
+
+	log.Info("Writing member announce", "cluster", r.GetClusterID())
+
+	memberAnnounceList := &multiclusterv1alpha1.MemberClusterAnnounceList{}
+	if err := r.List(context.TODO(), memberAnnounceList, client.InNamespace(r.GetNamespace())); err != nil {
+		return err
+	}
+	var localClusterMemberAnnounce multiclusterv1alpha1.MemberClusterAnnounce
+	localClusterMemberAnnounceExist := false
+	if len(memberAnnounceList.Items) != 0 {
+		for _, memberAnnounce := range memberAnnounceList.Items {
+			if strings.Compare(memberAnnounce.ClusterID, string(r.remoteClusterManager.GetLocalClusterID())) == 0 {
+				localClusterMemberAnnounceExist = true
+				localClusterMemberAnnounce = memberAnnounce
+			}
+		}
+	}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFunc()
+	if localClusterMemberAnnounceExist {
+		if r.remoteClusterManager.GetElectedLeaderClusterID() != common.INVALID_CLUSTER_ID {
```

Comment:
I think better to add more comments somewhere to explain how leader election is done. Not easy to figure out just from code.